### ISSUE TITLE
Missed optimization: loop evaluation can be folded into a constant

### ISSUE THREAD

**Issue by dcci** [swiftlang/swift#76535]
### Description

https://godbolt.org/z/KfW9nPG4j

Reduced from a large example.
This loop can be folded into a constant. Commenting this line (which is dead),

```
            // Dead nested while loop, will never be reached
            while (number + 1 == 5 && number * 2 == 8) {}
```

results in this code being generated:

```
output.sum() -> Swift.Int:
        mov     eax, 10
        ret
```

So, the dead code prevents this loop from being optimized.

### Reproduction

```swift -O
```

### Expected behavior

The loop is folded into a constant

### Environment

trunk/nightly x86-64

**Comment by jamieQ** [comment]
to clarify: in the example, when `number` is `4`, wouldn't the nested loop be expected to run forever?

**Comment by dcci** [comment]
> to clarify: in the example, when `number` is `4`, wouldn't the nested loop be expected to run forever?

That is correct, but I think you can slightly change the example keeping the codegen identical, e.g.

```
            // Dead nested while loop, will never be reached
            while (number + 1 == 5 && number * 2 == 9) {}
```
